reaktive/src/jvmJsCommonMain/kotlin/com/badoo/reaktive/utils/serializer/SerializerImpl.kt

```diff
@@ -1,76 +1,82 @@
 package com.badoo.reaktive.utils.serializer
 
+import com.badoo.reaktive.utils.atomic.AtomicInt
 import com.badoo.reaktive.utils.queue.Queue
+import kotlin.jvm.Volatile
 
-internal abstract class SerializerImpl<in T>(queue: Queue<T>) : Serializer<T> {
+/*
+ * Derived from RxJava SerializedEmitter
+ */
+internal abstract class SerializerImpl<in T>(
+    private val queue: Queue<T>
+) : Serializer<T> {
 
-    private var queue: Queue<T>? = queue
-    private val monitor = Any()
-    private var isDraining = false
+    @Volatile
+    private var isDone = false
+    private val counter = AtomicInt()
 
     override fun accept(value: T) {
-        val queueToDrain =
-            synchronized(monitor) {
-                val queue = queue ?: return
+        if (isDone) {
+            return
+        }
 
-                if (isDraining) {
-                    queue.offer(value)
-                    return
-                }
+        if (counter.compareAndSet(0, 1)) {
+            if (!onValue(value)) {
+                isDone = true
+                return
+            }
 
-                isDraining = true
-                queue
+            if (counter.addAndGet(-1) == 0) {
+                return
+            }
+        } else {
+            synchronized(queue) {
+                queue.offer(value)
             }
 
-        if (!processValue(value)) {
-            return
+            if (counter.addAndGet(1) > 1) {
+                return
+            }
         }
 
-        queueToDrain.drain()
+        drainLoop()
     }
 
     override fun clear() {
-        synchronized(monitor) {
-            queue?.clear()
-        }
+        synchronized(queue, queue::clear)
     }
 
-    private fun Queue<T>.drain() {
+    abstract fun onValue(value: T): Boolean
+
+    private fun drainLoop() {
+        var missed = 1
         while (true) {
-            val value =
-                synchronized(monitor) {
-                    if (isEmpty) {
-                        onDrainFinished(false)
-                        return
+            while (true) {
+                var isEmpty = false
+                var value: T? = null
+
+                synchronized(queue) {
+                    isEmpty = queue.isEmpty
+                    if (!isEmpty) {
+                        value = queue.poll()
                     }
+                }
 
-                    @Suppress("UNCHECKED_CAST")
-                    poll() as T
+                if (isEmpty) {
+                    break
                 }
 
-            if (!processValue(value)) {
-                return
+                @Suppress("UNCHECKED_CAST")
+                if (!onValue(value as T)) {
+                    isDone = true
+                    return
+                }
             }
-        }
-    }
 
-    private fun processValue(value: T): Boolean {
-        if (!onValue(value)) {
-            synchronized(monitor) {
-                onDrainFinished(true)
+            missed = counter.addAndGet(-missed)
```

Comment:
This is a trick to avoid `synchronized`. The idea is that we remember `counter` value at the beginning of `drain`. At the end we should check if the `counter` is still same value (current counter value minus remembered value is 0). If it is same then we can finish `drain`. Otherwise we should remember the new `counter` value and repeat the process. The `comparison` is atomic, so either the current thread will start another `drain` round or another thread will start `drain` if there is a race.

Reply:
Thank you :)